### PR TITLE
Query: Support per endpoint TLS configuration

### PR DESCRIPTION
**Pull request by Naman-B-Parlecha**
<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->

<!--
    Don't forget about CHANGELOG!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->
I m still working on this have some edge case and tests to add. Will update description soon.

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

- Related to: #8586
 ## Changes                              

  Add per-endpoint and default TLS configuration for the Query component via `endpoint.sd-config`.                                        
                                          
  - Each endpoint can now specify its own `client_config` with TLS settings, server name, and compression                                 
  - A `default_client_config` can be set in the YAML config to apply TLS settings to all endpoints that don't specify their own           
  - The existing `--grpc-client-tls-*` CLI flags are deprecated (hidden from `--help`) and will be removed after v0.43.0                
  - A deprecation warning is logged when deprecated CLI flags are used                                                                    
                                                                                                                                        
  ### Precedence order                                                                                                                    
  1. Per-endpoint `client_config` (highest)                                                                                               
  2. `default_client_config` from YAML                                                                                                    
  3. `--grpc-client-tls-*` CLI flags (deprecated, lowest)                                                                                 
                                                                                                                                        
  ### Example config                                                                                                                      
  
  ```yaml                                                                                                                                 
  default_client_config:
    tls_config:
      enabled: true
      insecure_skip_verify: true
      cert_file: "/path/to/client.crt" 
      key_file: "/path/to/client.key"    
      ca_file: "/path/to/ca.crt"    
      min_version: "1.3"
    compression: "snappy"         
                                             
  endpoints:                              
    - address: "store1:10901"
      client_config:                                                                                                                      
        tls_config:
          enabled: true                                                                                                                   
          cert_file: "/path/to/client.crt"                                                                                              
          key_file: "/path/to/client.key"                                                                                               
          ca_file: "/path/to/ca.crt"                                                                                                      
        server_name: "store"
        compression: "snappy"                                                                                                             
```
More [documentation](https://github.com/thanos-io/thanos/pull/8594/files#diff-04f662ee00835250889473f4b6dc2cd116ae5901755a17ab454f84abfb108c2f)

<!-- Enumerate changes you made -->

## Verification
Added debug logs to check if switching between global and per endpoint configuration works as intended
```sh
./thanos query --log.level=debug --endpoint.sd-config-file=test_per_endpoint.yml

ts=2025-12-20T17:21:32.579477401Z caller=endpointset.go:429 level=debug msg="configured endpoint" addr=localhost:10901 tls=global compression=none
ts=2025-12-20T17:21:32.579526715Z caller=endpointset.go:429 level=debug msg="configured endpoint" addr=localhost:10902 tls="global (fallback)" compression=none
ts=2025-12-20T17:21:32.57954552Z caller=endpointset.go:429 level=debug msg="configured endpoint" addr=localhost:10903 tls="global (fallback)" compression=none
ts=2025-12-20T17:21:32.579550379Z caller=endpointset.go:429 level=debug msg="configured endpoint" addr=localhost:10904 tls=global compression=none
ts=2025-12-20T17:21:32.579554076Z caller=endpointset.go:429 level=debug msg="configured endpoint" addr=localhost:10905 tls=global compression=none
ts=2025-12-20T17:21:32.590883522Z caller=endpointset.go:429 level=debug msg="configured endpoint" addr=localhost:10906 tls=per-endpoint compression=none
```
<!-- How you tested it? How do you know it works? -->

### CODE REVIEW

File: cmd/thanos/endpointset.go [hunk 1]
```diff
@@ -41,11 +41,26 @@ type fileContent interface {
...
 type endpointSettings struct {
-	Strict        bool   `yaml:"strict"`
-	Group         bool   `yaml:"group"`
-	Address       string `yaml:"address"`
-	ServiceConfig string `yaml:"service_config"`
+	Strict        bool             `yaml:"strict"`
+	Group         bool             `yaml:"group"`
+	Address       string           `yaml:"address"`
+	ServiceConfig string           `yaml:"service_config"`
+	ClientConfig  clientGRPCConfig `yaml:"grpc_client_config"`
```
Comment (MichaHoffmann) [inline cmd/thanos/endpointset.go]: just client_config maybe?

**Comment by Naman-B-Parlecha** [comment]
@MichaHoffmann i have refactored to `client_config` PTAL!!

File: cmd/thanos/endpointset.go [hunk 2]
```diff
@@ -41,11 +41,26 @@ type fileContent interface {
 	Path() string
 }
 
+type tlsConfig struct {
+	Enabled                  bool   `yaml:"enabled"`
+	InsecureSkipVerification bool   `yaml:"insecure_skip_verify"`
+	CertFile                 string `yaml:"cert_file"`
+	KeyFile                  string `yaml:"key_file"`
+	CAFile                   string `yaml:"ca_file"`
+}
```
Comment (ringerc) [inline cmd/thanos/endpointset.go:53]: Does it make sense to expose any other parts of https://pkg.go.dev/crypto/tls#Config here? 
I see that `ServerName` is in the containing `clientConfig` so doesn't need adding here.

Exposing `MinVersion` may be good, as it's always best to force the highest safe TLS version to limit downgrade attacks.

A docstring linking to the crypto/tls docs would likely make sense here too.

File: cmd/thanos/endpointset.go [hunk 3]
```diff
@@ -63,6 +78,24 @@ type endpointConfigProvider struct {
...
 
+func (cfg *clientConfig) UseGlobalTLSOpts() bool {
+	return !cfg.TLSConfig.Enabled &&
+		cfg.TLSConfig.CertFile == "" &&
+		cfg.TLSConfig.KeyFile == "" &&
+		cfg.TLSConfig.CAFile == "" &&
+		cfg.ServerName == ""
+}
+
+func (cfg *clientConfig) ValidateCompression() error {
```
Comment (ringerc) [inline cmd/thanos/endpointset.go]: Shouldn't this just be `ValidateConfig()`  or `ValidateClientConfig()` as there'll likely be other things to validate in the config later?

File: cmd/thanos/endpointset.go [hunk 4]
```diff
@@ -353,16 +392,49 @@ func setupEndpointSet(
...
+			var tlsOpt grpc.DialOption
+			useGlobalConfig := ecfg.ClientConfig.UseGlobalTLSOpts()
+			if useGlobalConfig {
+				tlsOpt = globalTLSOpt
+			} else if tlsEnabled {
+				var err error
+				// We configure endpoint level TLS first and check if it fails we use fallback global TLS Opts
+				tlsOpt, err = extgrpc.StoreClientTLSCredentials(logger, ecfg.ClientConfig.TLSConfig.Enabled, ecfg.ClientConfig.TLSConfig.InsecureSkipVerification, ecfg.ClientConfig.TLSConfig.CertFile, ecfg.ClientConfig.TLSConfig.KeyFile, ecfg.ClientConfig.TLSConfig.CAFile, ecfg.ClientConfig.ServerName)
+				if err != nil {
+					level.Error(logger).Log("msg", "failed to create endpoint TLS, falling back to global", "addr", addr, "err", err)
```
Comment (ringerc) [inline cmd/thanos/endpointset.go]: Is this a secure behaviour?

It might be desirable for some users, but I suggest it at least be something that can be disabled with a flag in the clientConfig to prevent fallback. I'd be inclined to block fallback by default, but permit it if someone sets a `allowFallbackToGlobal: true` in the config.

It seems like the intention here was probably to inherit the global tls cert, key, and CA paths so they don't have to be repeated in each endpoint config. This only makes sense to do if the global CLI config has `--no-grpc-client-tls-secure` but sets a `--grpc-client-tls-ca` etc. Then it's possible to set `enabled: true` on endpoints where TLS is desired, without repeating the certs each time.

But in that case it would make no sense to log at error level - or anything > debug - since this is normal and intended behaviour. 

And it'd IMO make more sense to check if each per-endpoint setting is blank, and if so, inherit the global setting. Rather than trying to create a config and falling back to global on failure, especially without checking the exact error that caused the failure.

I suggest using pointers for the fields within `type tlsConfig struct` so it is possible to differentiate between "unset, inherit global default" and "explicitly set to the empty string by the user".

Then when applying the configuration, merge-in global defaults for unset fields. This should probably be done in initial argument parsing so validation can take them into consideration.

File: docs/components/query.md [hunk 5]
```diff
@@ -291,6 +291,35 @@ In case of nested Thanos Query components, it's important to note that tenancy e
...
+    client_config:
+      tls_config:
+        enabled: <bool>
+        insecure_skip_verify: <bool>
+        cert_file: "<path>"
+        key_file: "<path>"
+        ca_file: "<path>"
+      server_name: "<str>"
+      compression: "<str>"  # "none" or "snappy"
+```
```
Comment (ringerc) [inline docs/components/query.md:326]: Suggest a link to crypto/tls docs for Configuration as a reference for the fields

File: docs/components/query.md [hunk 6]
```diff
@@ -291,6 +291,35 @@ In case of nested Thanos Query components, it's important to note that tenancy e
...
+
+When `enabled` is true, TLS is configured to make use of certs path provided for the endpoint. If no path/invalid path is provided, global TLS configuration is used as a fallback. If `enabled` is false, cleartext is used.
```
Comment (ringerc) [inline docs/components/query.md]: Not convinced by this behaviour. See https://github.com/thanos-io/thanos/pull/8594#discussion_r2761550150

**Review by ringerc — COMMENTED** [review]
Thanks for this. While I'm not a proper project member, I hope this initial review is helpful. I'll see if I can also test out the change as this is functionality I happen to need.

I note that there is a test failure from CI below, but it appears to be unrelated to this change.

**Comment by Naman-B-Parlecha** [comment]
Hey @ringerc! Thanks for the review, [comment](https://github.com/thanos-io/thanos/pull/8594#discussion_r2761550150) completely make sense i will try and refactor based on your suggestions

File: cmd/thanos/endpointset.go (hunk 3 above)
Comment (Naman-B-Parlecha) [reply cmd/thanos/endpointset.go]: No `ValidateCompression` cuz we are checking for Compression type and not the either config for per endpoint 

ValidateConfig would happen in - https://github.com/Naman-B-Parlecha/thanos/blob/NamanParlecha/PerEndpointTLS/cmd/thanos/endpointset.go#L147-L161

Not sure what i was thinking but  `ValidateCompression` should be a private method so refactoring to `validateCompression`

File: cmd/thanos/endpointset.go (hunk 4 above)
Comment (ringerc) [reply cmd/thanos/endpointset.go]: Looks resolved by new push.

File: docs/components/query.md (hunk 6 above)
Comment (ringerc) [reply docs/components/query.md]: Looks resolved by new push

**Comment by ringerc** [comment]
> Hey @ringerc! Thanks for the review, [comment](https://github.com/thanos-io/thanos/pull/8594#discussion_r2761550150) completely make sense i will try and refactor based on your suggestions

Nice, thanks. I'd started a local tree to do that but you got there much faster and better than I did.

I've pinged #thanos-dev on Slack

File: docs/components/query.md [hunk 7]
```diff
@@ -291,6 +291,36 @@ In case of nested Thanos Query components, it's important to note that tenancy e
...
+        insecure_skip_verify: <bool>
+        cert_file: "<path>"
+        key_file: "<path>"
+        ca_file: "<path>"
+      server_name: "<str>"
+      compression: "<str>"  # "none" or "snappy"
+```
+Reference https://pkg.go.dev/crypto/tls#Config for Configuring the fields
+
+When `enabled` is true, TLS is configured to make use of certs path provided for the endpoint. If no path/invalid path is provided, global TLS configuration is used as a fallback. If `enabled` is false, cleartext is used.
```
Comment (ringerc) [inline docs/components/query.md]: This documentation needs updating to reflect the changes.

````suggestion
TLS may be overridden on a per-endpoint basis when configuring endpoints using file-based service discovery.  This makes it possible to use TLS for some endpoints but not others, to use different TLS trust domains for different endpoints, and so on.

Each entry in the per-endpoint `client_config` overrides the corresponding settings from the global configuration provided by [CLI options](#flags). To explicitly unset a global setting for one endpoint, set it to the empty string in the per-endpoint configuration. To inherit it, omit the setting key entirely. Each setting is independent, so it is possible to (e.g.) override the per-endpoint client cert and key while inheriting the globally-configured CA cert.

Example file service-discovery config file in YAML, showing the corresponding [global flag](#flags) for each per-endpoint option as a comment:

```yaml
endpoints:
  - address: "<endpoint-address>"
    client_config:
      tls_config:
        enabled: <bool>              # --[no-]grpc-client-tls-secure
        insecure_skip_verify: <bool> # --[no-]grpc-client-tls-skip-verify
        cert_file: "<path>"          # --grpc-client-tls-cert
        key_file: "<path>"           # --grpc-client-tls-key
        ca_file: "<path>"            # --grpc-client-tls-ca
        min_version: "<str>"         # --grpc-server-tls-min-version
      server_name: "<str>"           # --grpc-client-server-name
      compression: "<str>"           # --grpc-compression "none" or "snappy"
```

Just like with CLI flags, it is possible to set a cert, key and CA without actually enabling TLS. This is not considered an error. Make sure to explicitly set `enabled: true` if TLS is desired for an endpoint.

There is no support for per-endpoint TLS overrides for endpoints supplied by DNS service discovery or static CLI options.

See also https://pkg.go.dev/crypto/tls#Config for lower-level field details.
````

**Review by ringerc — COMMENTED** [review]
I've left a suggestion for a docs update. Thanks very much for your changes.

**Comment by ringerc** [comment]
It looks like you may need to `make white-noise-cleanup` too, as the docs checks are complaining. (Assuming I read the CI check correctly.)

It doesn't like the whitespace around the cli flags edits.

**Comment by Naman-B-Parlecha** [comment]
Thanks @ringerc applied ur suggestion on the docs have applied them!!

The CI for Docs is broken and has alot of error not related to this PR :((

**Comment by ringerc** [comment]
@Naman-B-Parlecha Grab this commit and apply it. I couldn't work out how to raise a PR for you against this branch.


Edit: actually apply this with `git am 0001-Fix-docs-formatting.patch` instead: 
[0001-Fix-docs-formatting.patch](https://github.com/user-attachments/files/25088816/0001-Fix-docs-formatting.patch)


alternatively grab this commit https://github.com/thanos-io/thanos/commit/a39f3c24010167ad3f3d0e6a372655e017b0be52

I suspect the e2e tests are broken, but the docs tests just need this content patch.

**Comment by Naman-B-Parlecha** [comment]
@ringerc @MichaHoffmann we didn't have tests for TLS per endpoint and i was testing thing out by logging so added some mock test to replicate `setupEndpointSet`.

**Comment by ringerc** [comment]
@Naman-B-Parlecha Looks like there are merge conflicts now

**Comment by Naman-B-Parlecha** [comment]
@ringerc resolved!

**Comment by ringerc** [comment]
Docs check failure is unrelated to this PR

> error: 		docs/contributing/coding-style-guide.md:7: "https://www.dabapps.com/blog/cognitive-load-programming/" not accessible even after retry; status code 503: Service Unavailable

**Comment by ringerc** [comment]
@MichaHoffmann wrote the original file-sd code, maybe he'll have a chance to review this?

**Comment by ringerc** [comment]
I pinged the Thanos dev channel about this, but so far I think everyone who can review and approve is busy. I'll try to check back later. Consider joining the CNCF Slack and the Thanos Dev channel yourself if you can @Naman-B-Parlecha

**Comment by ringerc** [comment]
@MichaHoffmann As the original author of the endpoint file-based config, do you think there's any chance you might be able to look over this change?

File: cmd/thanos/endpointset_test.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+	"github.com/efficientgo/core/testutil"
+)
+
+func boolPtr(b bool) *bool    { return &b }
+func strPtr(s string) *string { return &s }
+
+func TestPerEndpointTLSConfig(t *testing.T) {
+	globalTLSConfig := &tlsConfig{
+		Enabled:  boolPtr(true),
```
Comment (MichaHoffmann) [inline cmd/thanos/endpointset_test.go]: can we use "new" here?

File: pkg/cacheutil/redis_client.go [hunk 9]
```diff
@@ -187,9 +187,9 @@ func NewRedisClientWithConfig(logger log.Logger, name string, config RedisClient
 	var tlsConfig *tls.Config
 	if config.TLSEnabled {
 		userTLSConfig := config.TLSConfig
-
+		// TODO(naman): pass min TLS version from config.
```
Comment (MichaHoffmann) [inline pkg/cacheutil/redis_client.go]: can we do that?

File: cmd/thanos/endpointset_test.go (hunk 8 above)
Comment (ringerc) [reply cmd/thanos/endpointset_test.go]: Like

```go
	globalTLSConfig := &tlsConfig{
		Enabled:  new(bool),
		CertFile: new(string),
		KeyFile:  new(string),
		CAFile:   new(string),
	}
	*globalTLSConfig.Enabled = true
	*globalTLSConfig.CertFile = "/mock-path.pem"
	*globalTLSConfig.KeyFile = "/mock-key-path.pem"
	*globalTLSConfig.CAFile = "/mock-ca-path.pem"
```

or

```go
	var (
		gtlsEnabled = true
		gtlsCertFile = "/mock-path.pem"
		gtlsKeyFile = "/mock-key-path.pem"
		gtlsCAFile = "/mock-ca-path.pem"
	)
	globalTLSConfig := &tlsConfig{
		Enabled:  &gtlsEnabled,
		CertFile: &gtlsCertFile,
		KeyFile:  &gtlsKeyFile,
		CAFile:   &gtlsCAFile,
	}
```

?

If so:
* It separates allocation and assignment, which IMO makes the test harness setup less clear
* it can get ugly fast in test code where you're often constructing large arrays of complex types in-place
* `new(...)` is only in Go 1.26, but according to `go.mod` Thanos only targets 1.25 currently.

Personally I tend to use the awful hack `&[]bool{true}[0]` 😆 but I don't like to use it in code other people will have to endure. 

```go
	globalTLSConfig := &tlsConfig{
		Enabled:  &[]bool{true}[0],
		CertFile: &[]string{"/mock-path.pem"}[0],
		KeyFile:  &[]string{"/mock-key-path.pem"}[0],
		CAFile:   &[]string{"/mock-ca-path.pem"}[0],
	}
```

As golang lacks the developer convenience of supporting taking the address of a literal by implicitly allocating a heap object for it, IMO `boolPtr`, `strPtr` seem less bad than tangling up the test setup to allocate then separately set values, or splitting them into a `var` section beforehand.

This also ensures that the values in each test don't get shared, potentially creating surprising results if bugs cause the target values to be written to.

Note that they're pointer holders in the config so the code can differentiate between "explicitly set to empty/false" and "not set".

File: pkg/cacheutil/redis_client.go (hunk 9 above)
Comment (ringerc) [reply pkg/cacheutil/redis_client.go]: The reason it's not done in this patch appears to be that there's no flag for the gRPC client minimum TLS version in Receive so there's nothing to pass. The existing flag `--remote-write.server-tls-min-version` is actually for the HTTP server. There _is_ a global `--grpc-client-tls-min-version` flag, but the `--grpc-client-tls-` options are only used in Query, not Receive.

The PR could add a `--remote-write.client-tls-min-version`, but that'd expand its scope from just per-endpoint client TLS for Query. I'm guessing that's why they were TODOd for now.

Comment (ringerc) [reply pkg/cacheutil/redis_client.go]: When looking at this I found a related issue. The docs for `--remote-write.server-tls-min-version` say it's for the gRPC server, but it's actually for the _HTTP_ server. Small fix PR here https://github.com/thanos-io/thanos/pull/8700

Comment (ringerc) [reply pkg/cacheutil/redis_client.go]: IMO these TODOs could be addressed with a follow-up PR. I've updated my branch https://github.com/thanos-io/thanos/compare/main...ringerc:thanos-patches:NamanParlecha/PerEndpointTLS to address them with:

* https://github.com/thanos-io/thanos/commit/f1d6bce96a529fc4be7c7caeb48f58fdf3e68fa4
* https://github.com/thanos-io/thanos/commit/967a085336403609d39f9cf05ba947447abbd014
* https://github.com/thanos-io/thanos/commit/6eedaf2cb1d7d0c5ca2c7d764e39a7ef0beee38b
* https://github.com/thanos-io/thanos/commit/f80b55a0562dd20ffc88eca469e2ce58c0d020cc

I'd be happy to rebase them on top of this PR once it is merged, and make a new PR for these changes.

But if you'd rather do it in one shot, they can be merged into this branch with:

```
git fetch origin
git rebase origin/main
git remote add ringerc https://github.com/ringerc/thanos-patches
git fetch ringerc NamanParlecha/PerEndpointTLS
git merge ringerc/NamanParlecha/PerEndpointTLS
```

or by `git am`'ing these patches in sequence:

* [0001-feat-Set-gRPC-client-min-TLS-vers-in-Receive.patch.txt](https://github.com/user-attachments/files/25730188/0001-feat-Set-gRPC-client-min-TLS-vers-in-Receive.patch.txt)
* [0002-chore-Function-to-check-if-a-TLS-version-string-is-v.patch.txt](https://github.com/user-attachments/files/25730190/0002-chore-Function-to-check-if-a-TLS-version-string-is-v.patch.txt)
* [0003-feat-support-min_version-in-Redis-TLS-config.patch](https://github.com/user-attachments/files/25730263/0003-feat-support-min_version-in-Redis-TLS-config.patch)
* [0004-chore-Set-TLS-min-version-in-TLS-test.patch.txt](https://github.com/user-attachments/files/25730192/0004-chore-Set-TLS-min-version-in-TLS-test.patch.txt)

**Comment by ringerc** [comment]
Note doc check is failing due to unrelated

> error: 		docs/contributing/coding-style-guide.md:7: "https://www.dabapps.com/blog/cognitive-load-programming/" not accessible even after retry; status code 503: Service Unavailable

Perhaps the docs checker should _not_ be fetching remote  URIs for unchanged files every run?

File: pkg/cacheutil/redis_client.go (hunk 9 above)
Comment (Naman-B-Parlecha) [reply pkg/cacheutil/redis_client.go]: Thanks @ringerc for picking this up in ur PR!!

**Comment by Naman-B-Parlecha** [comment]
> Note doc check is failing due to unrelated
> 
> > error: 		docs/contributing/coding-style-guide.md:7: "https://www.dabapps.com/blog/cognitive-load-programming/" not accessible even after retry; status code 503: Service Unavailable
> 
> Perhaps the docs checker should _not_ be fetching remote URIs for unchanged files every run?

That would be great but then if no one ever makes changes to doc - we will never know if links are outdate/broken??

**Comment by ringerc** [comment]
> That would be great but then if no one ever makes changes to doc - we will never know if links are outdate/broken??

Run a periodic, scheduled check off `main` to report on them. That's what I do elsewhere. Saves hammering 3rd party sites in PRs.

File: pkg/cacheutil/redis_client.go (hunk 9 above)
Comment (Naman-B-Parlecha) [reply pkg/cacheutil/redis_client.go]: I have rebased ur commits to the branch Thanks for helping out!!

**Comment by Naman-B-Parlecha** [comment]
@ringerc can u signoff your commits for the DCO to pass

**Comment by ringerc** [comment]
@Naman-B-Parlecha Done and pushed the amended commits to https://github.com/ringerc/thanos-patches/tree/NamanParlecha/PerEndpointTLS

Sorry I forgot to add the sign-off tag.

You can

```
git remote add ringerc	https://github.com/ringerc/thanos-patches
git fetch ringerc
git reset --hard remotes/ringerc/NamanParlecha/PerEndpointTLS
```

to replace your current tree entirely with mine; I've only changed the sign-offs and added one commit to fix a docs checker complaint. Otherwise, you can `git rebase -i rebase` away my 4 patches from your tree, then `git am` these replacements. (Github is such a pain for working on collaborative PRs.)

* [0001-feat-Set-gRPC-client-min-TLS-vers-in-Receive.patch](https://github.com/user-attachments/files/25752649/0001-feat-Set-gRPC-client-min-TLS-vers-in-Receive.patch)
* [0002-chore-Function-to-check-if-a-TLS-version-string-is-v.patch](https://github.com/user-attachments/files/25752650/0002-chore-Function-to-check-if-a-TLS-version-string-is-v.patch)
* [0003-feat-support-min_version-in-Redis-TLS-config.patch](https://github.com/user-attachments/files/25752651/0003-feat-support-min_version-in-Redis-TLS-config.patch)
* [0004-chore-Set-TLS-min-version-in-TLS-test.patch](https://github.com/user-attachments/files/25752652/0004-chore-Set-TLS-min-version-in-TLS-test.patch)
* [0005-docs-Update-TLS-config-docs-for-min_version-changes.patch](https://github.com/user-attachments/files/25752921/0005-docs-Update-TLS-config-docs-for-min_version-changes.patch)


You'll need to `git commit --amend -s` your top commit too, to add the sign-off.

File: cmd/thanos/endpointset_test.go (hunk 8 above)
Comment (Naman-B-Parlecha) [reply cmd/thanos/endpointset_test.go]: i think using boolPtr and strPtr is fine here? 
can i mark this as resolved @MichaHoffmann and @ringerc?

File: cmd/thanos/config.go [hunk 10]
```diff
@@ -81,18 +82,17 @@ func (gc *grpcClientConfig) registerFlag(cmd extkingpin.FlagClause) *grpcClientC
 	cmd.Flag("grpc-client-server-name", "Server name to verify the hostname on the returned gRPC certificates. See https://tools.ietf.org/html/rfc4366#section-3.1").Default("").StringVar(&gc.serverName)
 	compressionOptions := strings.Join([]string{snappy.Name, compressionNone}, ", ")
 	cmd.Flag("grpc-compression", "Compression algorithm to use for gRPC requests to other clients. Must be one of: "+compressionOptions).Default(compressionNone).EnumVar(&gc.compression, snappy.Name, compressionNone)
-
+	cmd.Flag("grpc-client-tls-min-version",
```
Comment (GiedriusS) [inline cmd/thanos/config.go:95]: The knowledge about what TLS versions are allowed is scattered across multiple places. Maybe we could either and or: 
- Create some "global" array with allowed versions that will be referred to from other places
- Use EnumVar() here to check against that array
- Add a custom type with some function that does validation and/or a constructor

File: cmd/thanos/endpointset.go [hunk 11]
```diff
@@ -63,6 +81,91 @@ type endpointConfigProvider struct {
...
+// applyDefaults sets default values for nil pointer fields.
+// If fallback is provided, uses those values before falling back to defaults.
+// Call this only after UseGlobalTLSOpts() check, otherwise that check will always return false.
+func (tc *tlsConfig) applyDefaults(fallback *tlsConfig) {
+	if tc.Enabled == nil {
+		// If user provided any TLS config (cert/key/CA/insecure_skip_verify),
+		// we assume they want TLS enabled rather than using global's enabled value
+		if tc.hasTLSConfig() {
+			defaultVal := true
+			tc.Enabled = &defaultVal
```
Comment (GiedriusS) [inline cmd/thanos/endpointset.go]: I think we can now use `new(true)`?

File: cmd/thanos/endpointset_test.go (hunk 8 above)
Comment (GiedriusS) [reply cmd/thanos/endpointset_test.go]: I think we can just upgrade to Go 1.26 use `new(...)`. That's a lot of boilerplate that we could avoid just by upgrading. Let me try that.

File: cmd/thanos/endpointset.go [hunk 12]
```diff
@@ -63,6 +81,91 @@ type endpointConfigProvider struct {
...
+		} else if fallback != nil && fallback.Enabled != nil {
+			tc.Enabled = fallback.Enabled
+		} else {
+			defaultVal := false
+			tc.Enabled = &defaultVal
+		}
+	}
+	if tc.InsecureSkipVerification == nil {
```
Comment (GiedriusS) [inline cmd/thanos/endpointset.go:116]: Maybe the boilerplate could be reduced by a function like this https://go.dev/play/p/YXjiVmGmqZ5?

**Comment by GiedriusS** [comment]
Updated to 1.26 here https://github.com/thanos-io/thanos/pull/8717 so we can remove lots of boilerplate.

**Comment by Naman-B-Parlecha** [comment]
@GiedriusS refactored all changed requested PLAT!! thanks:))

**Review by GiedriusS — COMMENTED** [review]
Could we mark --endpoint and other parameters as deprecated in cmd/thanos/query.go in this same PR? In other words, let's only support using files for setting endpoints. In next_version+1 we can remove them entirely.

**Comment by Naman-B-Parlecha** [comment]
> Could we mark --endpoint and other parameters as deprecated in cmd/thanos/query.go in this same PR? In other words, let's only support using files for setting endpoints. In next_version+1 we can remove them entirely.

yes read the [slack thread](https://cloud-native.slack.com/archives/CL25937SP/p1773888095921659) regarding this
will refactor!!

thanks

File: pkg/tls/options.go [hunk 13]
```diff
@@ -230,3 +247,9 @@ func getTlsVersion(tlsMinVersion string) (uint16, error) {
 
 	return validOption.tlsOption[tlsMinVersion], nil
 }
+
+func ValidateTlsVersion(tlsVersion string) error {
+
```
Comment (GiedriusS) [inline pkg/tls/options.go]: Spurious space and very shallow. We typically try to avoid such small functions as per the coding style. Could we just expose getTlsVersion instead?

Comment (ringerc) [reply pkg/tls/options.go]: I think that was my addition. Sounds reasonable.

**Comment by ringerc** [comment]
This is still waiting. What can we do to progress this to merge?

**Comment by ringerc** [comment]
@Naman-B-Parlecha This now needs rebase. Sorry it's taking so long, I'm trying to encourage this to get through final review.

**Comment by Naman-B-Parlecha** [comment]
The docs and unit test failures are not related to pr

**Review by GiedriusS — APPROVED** [review]

**Comment by ringerc** [comment]
Thanks very much @GiedriusS
